#!/usr/bin/env Rscript
# Takes as input a sample name to create the PyClone input tsv file
# using the clonality_mutation_table.txt file from strelka and segments.txt file generated by sequenza
# run the script inside the sequenza directory for each sample

DoSomatic2PyClone <- function() {
  mut.tab=read.table("clonality_mutation_table.txt",head=T,sep="\t")
  seg.cn=read.table("segments.txt",head=T)
  seq2PyClone<-somatic2PyClone(mut.tab, seg.cn,norm.cn = 2)
  write.table(seq2PyClone, file=("PyClone_output.txt"),col.names=T,row.names=F,sep="\t",quote=F)
}

somatic2PyClone <- function(mut.tab, seg.cn, norm.cn = 2) {
  mut.tab <- cbind(mut.tab[,c('Chr', 'Start', 'Ref','Alt', 'var_T', 'ref_T', 'Gene_refGene')], CNt = NA, A = NA, B = NA)
  for (i in 1:nrow(seg.cn)) {
    pos.filt <- mut.tab$Chr == seg.cn$chromosome[i] & mut.tab$Start >= seg.cn$start.pos[i] & mut.tab$Start <= seg.cn$end.pos[i] & seg.cn$A[i]>0
    mut.tab[pos.filt, c("CNt", "A", "B")] <- seg.cn[i, c("CNt", "A", "B")]
  }
  id          <- paste(mut.tab$Gene_refGene, mut.tab$Chr, mut.tab$Start, sep = ':')
  var.counts  <- mut.tab$var_T
  #var.counts  <- round(mut.tab$good.reads * mut.tab$F, 0)
  nor.counts  <- mut.tab$ref_T
  pyclone.tsv <- data.frame(mutation_id = id, 
                            ref_counts = nor.counts, 
                            var_counts = var.counts,
                            normal_cn = norm.cn,
                            minor_cn = mut.tab$B,
                            major_cn = mut.tab$A,
                            variant_freq = round(var.counts/(var.counts+nor.counts),2),	
                            genotype = paste(mut.tab$Ref,mut.tab$Alt,sep='>'))
  na.exclude(pyclone.tsv)
}

DoSomatic2PyClone()

